### ISSUE TITLE
Feature request: Relay runc logs to podman

### ISSUE THREAD

**Issue by filbranden** [containers/libpod#2285]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind feature

**Description**

When passing podman a `--log-level debug` or similar setting, it would be great if it could collect the `runc` logs and show them together with the podman logs.

**Steps to work around the issue:**

It is possible to collect `runc` logs, but that needs a wrapper and logs the runs to separate files (so not really good for finding correlations.)

For example, create `/usr/bin/logged_runc.sh` with contents:

```sh
#!/bin/sh
exec /usr/bin/runc --log "/tmp/runc.log-$$-$(date +%Y%m%dT%H%M%S)" "$@"
```

And then run:

```
# podman --log-level debug --runtime /usr/bin/logged_runc.sh run -t fedora:28 echo "Hello world!"
```

**Possible implementation:**

Perhaps a good way to implement this would be to open a pipe, clear FD_CLOEXEC on its writing end and pass `runc` a `--log /dev/fd/X` argument to write logs to that pipe. (At least, this is what I'd do if this was C.)

In podman, have a go routine read from that pipe and relay that message through the native logging system.

This assumes this podman process (with the go routine reading from the pipe) will be around for the lifetime of runc, not sure that's the case and there's such a thing.

I'm happy to contribute code, if needed. But I'd like some pointers on whether this is a good approach and which pitfalls to look for.

**Output of `podman version`:**

```
Version:            1.0.1-dev
RemoteAPI Version:  1
Go Version:         go1.12beta2
Git Commit:         "c9e1f365e294135968bdd4d754dca7e1044587dc-dirty"
Built:              Thu Jan 24 11:06:57 2019
OS/Arch:            linux/amd64
```

**Output of `podman info`:**

```
host:
  BuildahVersion: 1.6-dev
  Conmon:
    package: podman-1.0.1-17.dev.gitd5593b8.fc30.x86_64
    path: /usr/libexec/podman/conmon
    version: 'conmon version 1.12.0-dev, commit: d3cac30f4e81084c7bc36922e02817841ccc3456'
  Distribution:
    distribution: fedora
    version: "30"
  MemFree: 888102912
  MemTotal: 8360546304
  OCIRuntime:
    package: runc-1.0.0-70.dev.gite4fa8a4.fc30.x86_64
    path: /usr/bin/runc
    version: |-
      runc version 1.0.0-rc6+dev
      commit: 5b54487295859772ddfc57918f10e82994e501ae
      spec: 1.0.1-dev
  SwapFree: 2146942976
  SwapTotal: 2147479552
  arch: amd64
  cpus: 8
  hostname: rawhide.libvirt
  kernel: 5.0.0-0.rc3.git0.1.fc30.x86_64
  os: linux
  rootless: true
  uptime: 291h 32m 50.81s (Approximately 12.12 days)
insecure registries:
  registries: []
registries:
  registries:
  - docker.io
  - registry.fedoraproject.org
  - quay.io
  - registry.access.redhat.com
  - registry.centos.org
store:
  ConfigFile: /home/filbranden/.config/containers/storage.conf
  ContainerStore:
    number: 0
  GraphDriverName: overlay
  GraphOptions:
  - overlay.mount_program=/usr/bin/fuse-overlayfs
  GraphRoot: /home/filbranden/.local/share/containers/storage
  GraphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  ImageStore:
    number: 1
  RunRoot: /run/user/5192
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

QEMU-KVM, libvirt.

**Comment by mheon** [comment]
We have Conmon holding open stdin/out/err on the runc process, so we do have a monitor process that could potentially collect the logs. I don't know whether we'd want to force runc debug logging with --log-level=debug, though, it would make our lives a lot harder when tracking down pure Podman issues. I wouldn't mind adding this, but I think it would need to be an opt-in flag to enable runtime debugging.

(There's also the issues that I'm fairly certain the debug flag is runc-specific, and could potentially impact compatibility with other runtimes)

**Comment by filbranden** [comment]
Separate flag is just fine: `--runtime-log-level` perhaps?

I can give you more context on what I'm doing and how I'm bumping into issues with runc (so you can even reproduce this if you like.)

runc doesn't support cgroupv2 only (this is what I'm working to fix right now), so booting the system with `systemd.unified_cgroup_hierarchy=1` will make runc break.

After seeing podman fail, I re-ran it with `--log-level debug` to see if I found out more, but the messages are not useful to find out what went wrong:

```
...
DEBU[0000] Created OCI spec for container a9f043806b6e18a3deba541061d173c62166ef05d867d2de2942ad994d2f2870 at /var/lib/containers/storage/overlay-containers/a9f043806b6e18a3deba541061d173c62166ef05d867d2de2942ad994d2f2870/userdata/config.json
DEBU[0000] /usr/libexec/podman/conmon messages will be logged to syslog
DEBU[0000] running conmon: /usr/libexec/podman/conmon    args=[-s -c a9f043806b6e18a3deba541061d173c62166ef05d867d2de2942ad994d2f2870 -u a9f043806b6e18a3deba541061d173c62166ef05d867d2de2942ad994d2f2870 -r /usr/bin/logged_runc.sh -b /var/lib/containers/storage/overlay-containers/a9f043806b6e18a3deba541061d173c62166ef05d867d2de2942ad994d2f2870/userdata -p /var/run/containers/storage/overlay-containers/a9f043806b6e18a3deba541061d173c62166ef05d867d2de2942ad994d2f2870/userdata/pidfile -l /var/lib/containers/storage/overlay-containers/a9f043806b6e18a3deba541061d173c62166ef05d867d2de2942ad994d2f2870/userdata/ctr.log --exit-dir /var/run/libpod/exits --socket-dir-path /var/run/libpod/socket -t --log-level debug --syslog]
INFO[0000] Running conmon under slice machine.slice and unitName libpod-conmon-a9f043806b6e18a3deba541061d173c62166ef05d867d2de2942ad994d2f2870.scope
DEBU[0000] Cleaning up container a9f043806b6e18a3deba541061d173c62166ef05d867d2de2942ad994d2f2870
DEBU[0000] Tearing down network namespace at /var/run/netns/cni-297bef07-f095-3731-c852-139bfb5e3ad4 for container a9f043806b6e18a3deba541061d173c62166ef05d867d2de2942ad994d2f2870
INFO[0000] Got pod network &{Name:zealous_goldstine Namespace:zealous_goldstine ID:a9f043806b6e18a3deba541061d173c62166ef05d867d2de2942ad994d2f2870 NetNS:/var/run/netns/cni-297bef07-f095-3731-c852-139bfb5e3ad4 PortMappings:[] Networks:[] NetworkConfig:map[]}
INFO[0000] About to del CNI network podman (type=bridge)
DEBU[0000] unmounted container "a9f043806b6e18a3deba541061d173c62166ef05d867d2de2942ad994d2f2870"
ERRO[0000] error reading container (probably exited) json message: EOF
```

Using the wrapper with a logfile shows what's wrong:

```
$ cat /tmp/runc.log-978-20190206T185354
time="2019-02-06T18:53:54-08:00" level=warning msg="signal: killed"
time="2019-02-06T18:53:54-08:00" level=error msg="container_linux.go:344: starting container process caused \"process_linux.go:275: applying cgroup configuration for process caused \\"mountpoint for devices not found\\"\"
"
```

I really think showing all the logs merged into a single stream would have been best, since then the correlation becomes obvious. Perhaps, for clarity, it would be helpful to show the runc logs with a prefix such as `runc:` or `runc[PID]:`, so it's really clear those are not podman's...

So, again, separate flag is probably fine, but I still think it's useful to be able to enable debug on everything together and merge all the logs into a single stream, then no matter where things go wrong, they'll be captured...

Cheers!
Filipe

**Comment by rhatdan** [comment]
It could also be a libpod.conf option. That says whether to pass the debug option if set to the OCI runtime.

**Comment by mheon** [comment]
I was thinking the same. Maybe a more generic "runtime flags" option to set custom flags on all runtime invocations?

That doesn't get us the ability to log the actual messages as they occur, but it would allow making logfiles.

**Comment by giuseppe** [comment]
I agree on a generic "runtime flags", as probably `--log` is not part of the OCI runtime, so it will break other runtimes.

I've opened a PR for conmon to accept arguments to pass to the OCI runtime:

https://github.com/kubernetes-sigs/cri-o/pull/2062

once that lands, we can add support for additional runtime args to Podman

**Comment by giuseppe** [comment]
an interesting hack that is possible with this feature, is that an user can specify:

`--runtime-arg --debug --runtime-arg --log=/dev/stdout`

then gets the runtime debug log with:

`podman logs $CONTAINER`

**Comment by filbranden** [comment]
> an interesting hack that is possible with this feature, is that an user can specify:
>
> `--runtime-arg --debug --runtime-arg --log=/dev/stdout`
>
> then gets the runtime debug log with:
>
> `podman logs $CONTAINER`

I can definitely live with that :-)

One downside of specifying an external log file (e.g. `--log /tmp/runc.log`) is that `podman` will call `runc` twice (at least in my experiment) so I'm concerned about the log file being overwritten at some point... (Haven't really looked at whether the log file is opened for append or whether it's truncated...) The `/dev/stdout` "hack" looks good enough for me.

Thanks for looking into this @giuseppe !

**Comment by rhatdan** [comment]
@giuseppe @filbranden Are we done with this one?

**Comment by giuseppe** [comment]
@rhatdan the PR is stuck with the CI: https://github.com/kubernetes-sigs/cri-o/pull/2062

**Comment by rhatdan** [comment]
I just merged that PR>

**Comment by giuseppe** [comment]
@filbranden is the current fix in conmon good for you?  Can we close this issue? :-)

**Comment by filbranden** [comment]
Sorry I haven't had time to look into it (and next few days look very busy.)

Let's do this: I'll close this Issue now and if this comes up again in the future I'll reopen (or send a PR.) Sounds right?

Cheers!
Filipe

**Comment by srd424** [comment]
OK, I have crun issue I'm trying to debug, and trying to use the feature added by the PR here, but I'm clearly undercaffeinated today - how do I get these options to conmon from podman?

**Comment by giuseppe** [comment]
If you specify something like `podman run --runtime-arg ARG0 --runtime-arg ARG1` then ARG0 and ARG1 will be passed to the OCI runtime.  So you can do: `--runtime-arg --log=/a/path --runtime-arg --debug` and have `--log=/a/path` and `--debug` passed down to the OCI runtime.
